python/tutorials/gluon/01-attention-forward.py

```diff
@@ -719,203 +468,397 @@ def _mul_f32x2(a, b):
 
     for start_n in range(lo, hi, config.BLOCK_N):
-        qk, qk_consumer = qk_consumer.get(config.qk_layout)
-        if config.HEAD_DIM == 128:
-            p_tmem, p_bar, p_producer = p_producer.acquire()
+        s_tmem, s_bar, s_consumer = s_consumer.acquire()
+        qk = s_tmem.load(config.qk_layout)
 
         if STAGE == 2:
             # Prevent LLVM from hoisting the partial sums, which triggers spilling.
             offs_n = gl.inline_asm_elementwise("mov.b32 $0, $0;", "=r,r", [offs_n], dtype=gl.int32, is_pure=True,
                                                pack=1)
-            mask = offs_m[:, None] >= (start_n + offs_n[None, :])
-            qk = gl.where(mask, qk, -1.0e8)
+            mask = offs_m[:, None] < (start_n + offs_n[None, :])
+            qk = gl.where(mask, -1.0e8, qk)
         m_ij = gl.maximum(m_i, gl.max(qk, 1) * config.qk_scale)
-        if config.mi_use_tmem:
-            mi_producer = mi_producer.emplace(gl.convert_layout(m_ij.expand_dims(1), config.mi_2d_layout))
-        else:
-            mi_producer = mi_producer.emplace(m_ij)
-        qk = qk * config.qk_scale - m_ij[:, None]
-
-        if config.HEAD_DIM == 64:
-            p = gl.exp2(qk)
-            l_ij = gl.sum(p, 1)
-            alpha = gl.exp2(m_i - m_ij)
-            p_producer = p_producer.emplace(p.to(config.dtype))
-        else:
-            qk0, qk1, = qk.reshape([config.SPLIT_M, 2, config.BLOCK_N // 2]).permute(0, 2, 1).split()
-            p0 = gl.exp2(qk0)
-            p_tmem.slice(0, config.BLOCK_N // 2).store(p0.to(config.dtype))
-            p1 = gl.exp2(qk1)
-            p_tmem.slice(config.BLOCK_N // 2, config.BLOCK_N // 2).store(p1.to(config.dtype))
-            mbarrier.arrive(p_bar, count=1)
-            p = gl.join(p0, p1).permute(0, 2, 1).reshape([config.SPLIT_M, config.BLOCK_N])
-            p = gl.convert_layout(p, config.qk_layout)
-
-            l_ij = gl.sum(p, 1)
-            alpha = gl.exp2(m_i - m_ij)
+        alpha = gl.exp2(m_i - m_ij)
+
+        alpha_tmem = _borrow_s_as_alpha(config, s_tmem)
+        alpha_tmem.store(gl.convert_layout(alpha.expand_dims(1), config.alpha_2d_layout))
+        mbarrier.arrive(corr_bar, count=1)
+
+        qk = _mul_f32x2(qk, gl.full(config.qk_shape, config.qk_scale, gl.float32, qk.type.layout))
+        qk = _add_f32x2(qk, -m_ij[:, None])
+        qk0, qk1, = qk.reshape([config.SPLIT_M, 2, config.BLOCK_N // 2]).permute(0, 2, 1).split()
+
+        p_tmem = _borrow_s_as_p(config, s_tmem)
+        p0 = gl.exp2(qk0)
+        p_tmem.slice(0, config.BLOCK_N // 2).store(p0.to(config.dtype))
+        p1 = gl.exp2(qk1)
+        p_tmem.slice(config.BLOCK_N // 2, config.BLOCK_N // 2).store(p1.to(config.dtype))
+        mbarrier.arrive(s_bar, count=1)
 
+        _, corr_bar, corr_producer = corr_producer.acquire()
+
+        p = gl.join(p0, p1).permute(0, 2, 1).reshape([config.SPLIT_M, config.BLOCK_N])
+        p = gl.convert_layout(p, config.qk_layout)
```

Comment:
This shouldn't be needed any more after I introduced the slice layout for split, right?